icm42688: add emulator and tests

yperess opened the pull request:
- Add a register level emulator for the icm42688
- Fix a bug in the temperature calculation for the icm42688
- Add tests for data ready register and temperature

Commit:
emul: Add emulator for icm426888

Add an implementation for the icm42688 emulator that supports reading
any registers.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
icm42688: Fix temperature calculation

Fix the calculation which was adding the 25 degree offset too late.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
icm42688: Add test for data read and temperature regisers

Add tests that verify the behavior of the REG_INT_STATUS, REG_TEMP_DATA1,
and REG_TEMP_DATA0 registers

Signed-off-by: Yuval Peress <peress@google.com>

asemjonovs commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -359,15 +359,13 @@ static inline void icm42688_temp_c(int32_t in, int32_t *out_c, uint32_t *out_uc)
 
-	int64_t in100 = in * 100;
+	/* Shift by 25 degrees Celsius */
+	int64_t in100 = in * 100 + 25 * sensitivity;
```
Nit: Add parenthesis to help with order of operation.

yperess replied:
Done

aaronemassey commented on tests/drivers/sensor/icm42688/src/main.c:
```diff
@@ -0,0 +1,68 @@
+	 *   reg_val / 132.48 + 25
+	 */
+	temperature = ((22 - 25) * 13248) / 100;
+	printk("temperature = 0x%x (%d)\n", temperature, temperature);
```
Missed a print statement?

yperess replied:
Done

yperess commented on subsys/emul/sensor/icm42688/Kconfig:
```diff
@@ -0,0 +1,10 @@
+	depends on EMUL
+	help
+	  Enable the hardware emulator for the ICM42688. Doing so allows exercising
+	  sensor APIs for this IMU in native_posix.
```
Done

MaureenHelm commented on subsys/emul/CMakeLists.txt:
```diff
@@ -9,3 +9,4 @@ zephyr_library_sources_ifdef(CONFIG_EMUL_BMI160		emul_bmi160.c)
 
 add_subdirectory(i2c)
 add_subdirectory(espi)
+add_subdirectory(sensor)
```
Can you clean up the bmi160 emul build handling to be consistent with this pattern?

yperess replied:
I actually had a thought yesterday and wanted to get your take on this (as well as @keith-zephyr). I'm thinking of moving the emulators to be in the same directory as the drivers they emulate, it would make writing and updating them much easier.

MaureenHelm commented on subsys/emul/sensor/icm42688/include/zephyr/emul/sensor/icm42688/icm42688_emul.h:
```diff
@@ -0,0 +1,32 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_SUBSYS_EMUL_SENSOR_ICM42688_INCLUDE_ZEPHYR_EMUL_SENSOR_ICM42688_ICM42688_EMUL_H
+#define ZEPHYR_SUBSYS_EMUL_SENSOR_ICM42688_INCLUDE_ZEPHYR_EMUL_SENSOR_ICM42688_ICM42688_EMUL_H
```
OMG that is a long macro name. Why do you have `emul/sensor/icm42688` in the path twice?

MaureenHelm commented on subsys/emul/sensor/icm42688/src/icm42688_emul.c:
```diff
@@ -0,0 +1,98 @@
+	return 0;
+}
+
+static struct spi_emul_api icm42688_emul_spi_api = {
```
`const`?

MaureenHelm commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -359,15 +359,13 @@ static inline void icm42688_temp_c(int32_t in, int32_t *out_c, uint32_t *out_uc)
 	int64_t sensitivity = 13248; /* value equivalent for x100 1c */
 
-	int64_t in100 = in * 100;
+	/* Shift by 25 degrees Celsius */
```
```suggestion
	/* Offset by 25 degrees Celsius */
```

MaureenHelm commented on tests/drivers/sensor/icm42688/boards/native_posix.conf:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_EMUL=y
+CONFIG_SPI=y
+CONFIG_SENSOR=y
+CONFIG_ICM42688=y
```
I don't think CONFIG_ICM42688 or CONFIG_SPI need to be set here

Commit:
bmi160: move emulator to driver directory

Allow the emulator to sit in the same directory as the driver
implementation. This will make working on the emulator much easier and
keep the emulator subsystem directory clean by allowing it to focus on
the actual subsystem and not the use cases of the subsystem.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
emul: Cleanup emul directory

Avoid implementations in the subsys emul directory to keep the directory
focused on emulator subsystem instead of consumers. Consumers should be
implemented side-by-side to their drivers

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
test: verify accel/gyro data for icm42688

Verify the conversion of register values in both accel and gyro to
sensor_value for the icm42688.

Signed-off-by: Yuval Peress <peress@google.com>

teburd commented on drivers/sensor/sbs_gauge/emul_sbs_gauge.c:
```diff
@@ -21,7 +18,7 @@ LOG_MODULE_REGISTER(sbs_sbs_gauge);
 #include <zephyr/sys/byteorder.h>
 
-#include <sbs_gauge.h>
+#include "../../fuel_gauge/sbs_gauge/sbs_gauge.h"
```
I'm a bit confused by the include path and path to the .c file here, is this a fuel gauge or sensor emulated device? Both?

yperess replied:
Sorry, my IDE did this when I moved the file and I didn't catch it. Fixed

alevkoy commented on drivers/sensor/icm42688/icm42688_emul.c:
```diff
@@ -0,0 +1,98 @@
+	.io = icm42688_emul_io_spi,
+};
+
+#define ICM42688_EMUL_DATA(n) static struct icm42688_emul_data icm42688_emul_data_##n
```
Nit: Why have these one-line macros? The `icm42688_emul_cfg` definition looks fine as is.

yperess replied:
I expect this to grow as I add interrupt handling.

alevkoy commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -359,15 +359,13 @@ static inline void icm42688_temp_c(int32_t in, int32_t *out_c, uint32_t *out_uc)
-
-	/* Shift whole celsius 25 degress */
-	*out_c = *out_c + 25;
+	*out_uc = ((llabs(in100) - (llabs(*out_c)) * sensitivity) * INT64_C(1000000)) / sensitivity;
```
Nit: For me, the outer parens in `(llabs(*out_c))` make this line harder to parse than it would be without them.

alevkoy commented on tests/drivers/sensor/icm42688/boards/native_posix.conf:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_EMUL=y
+CONFIG_SENSOR=y
+CONFIG_EMUL_ICM42688=y
```
Wouldn't you need these configs no matter what board you ran this test on? Shouldn't they then be in `prj.conf`?

MaureenHelm commented on drivers/sensor/bmi160/Kconfig:
```diff
@@ -211,4 +211,16 @@ config BMI160_GYRO_ODR_3200
 
+
+config EMUL_BMI160
+	bool "Emulate a Bosch BMI160 accelerometer"
```
`depends on EMUL` ?

yperess replied:
Thanks

MaureenHelm commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -294,7 +294,7 @@ static inline void icm42688_accel_ms(struct icm42688_cfg *cfg, int32_t in, int32
 
 	/* micrometers/s^2 */
-	*out_ums = ((llabs(in_ms) - (llabs(*out_ms) * sensitivity * 1000000LL)) / sensitivity);
+	*out_ums = (in_ms - (*out_ms * sensitivity * 1000000LL)) / sensitivity;
```
Is this change related to `test: verify accel/gyro data for icm42688`?

yperess replied:
Yes, it was incorrect to take the diff of the abs

MaureenHelm commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -343,8 +343,8 @@ static inline void icm42688_gyro_rads(struct icm42688_cfg *cfg, int32_t in, int3
 	/* microrad/s */
-	*out_urads = ((llabs(in10_rads) - (llabs((*out_rads)) * sensitivity * 180LL * 1000000LL))) /
-		     (sensitivity * 180LL);
+	*out_urads =
+		(in10_rads - (*out_rads * sensitivity * 180LL * 1000000LL)) / (sensitivity * 180LL);
```
Is this change related to `test: verify accel/gyro data for icm42688`?

yperess replied:
Yes, it was incorrect to take the diff of the abs

MaureenHelm commented on include/zephyr/drivers/sensor.h:
```diff
@@ -640,7 +640,7 @@ static inline void sensor_g_to_ms2(int32_t g, struct sensor_value *ms2)
 static inline int32_t sensor_ms2_to_ug(const struct sensor_value *ms2)
 {
-	int64_t micro_ms2 = ms2->val1 * 1000000LL + ms2->val2;
+	int64_t micro_ms2 = (ms2->val1 * INT64_C(1000000)) + ms2->val2;
```
Is this change related to `test: verify accel/gyro data for icm42688`?

yperess replied:
Yes, `LL` wasn't casting correctly and I wanted to be more explicit on both the type and the order of operations.

teburd reviewed (dismissed):
It's nice to have tests covering both the fetch and conversion functions here. Clearly showed some errors, nice work.

yperess commented:
@teburd I need one more pass at review, there was a change to the emulation API that I needed to fix and I found a bug when using a negative temperature so I added a test for that too.

yperess commented:
My IDE snuck in a copyright change, need another review pass

teburd commented on drivers/sensor/icm42688/icm42688_emul.c:
```diff
@@ -0,0 +1,96 @@
+
+LOG_MODULE_DECLARE(ICM42688, CONFIG_SENSOR_LOG_LEVEL);
+
+#define NUM_REGS (UINT8_MAX >> 1)
```
Sort of :-)

The part has register banks (tdk/invensense seem to do this a lot) which can be selected by writing to an always available register. This will likely need to be updated when/if mocking that sort of pattern is ever needed. Clearly works fine for now though, not a blocker.

yperess replied:
Thanks, I'm working on the interrupt support for the driver and emulator now so I'll try to squeeze this in there.

teburd reviewed (dismissed):
Looks good, and its nice to have some bugs fixed

yperess commented:
@MaureenHelm @teburd I'm sorry, accidentally pushed the wrong branch, you can check the compare, there was no delta :cry:

yperess commented:
@stephanosio can this merge? It includes bug fixes for the ICM42688 driver along with tests for those fixes.

henrikbrixandersen commented on drivers/eeprom/CMakeLists.txt:
```diff
@@ -13,3 +13,5 @@ zephyr_library_sources_ifdef(CONFIG_EEPROM_EMULATOR eeprom_emulator.c)
 zephyr_library_sources_ifdef(CONFIG_EEPROM_XEC eeprom_mchp_xec.c)
 zephyr_library_sources_ifdef(CONFIG_EEPROM_FAKE eeprom_fake.c)
+
+zephyr_library_sources_ifdef(CONFIG_EMUL_EEPROM_AT2X emul_atmel_at24.c)
```
This breaks the existing naming scheme for EEPROM driver implementation files.

yperess replied:
This is the current name for the Kconfig so I was trying to do a minimal delta, I went ahead and included a rename and deprecation of the old Kconfig so we can move it to the correct naming scheme.

henrikbrixandersen replied:
This is still not addressed.

yperess replied:
Sorry, I thought the issue was the Kconfig.

henrikbrixandersen commented on drivers/eeprom/Kconfig:
```diff
@@ -41,6 +41,15 @@ config EEPROM_AT2X
 	  Enable support for Atmel AT2x (and compatible) I2C/SPI
 	  EEPROMs.
 
+config EMUL_EEPROM_AT2X
```
This breaks the existing naming scheme for EEPROM driver Kconfig options.

yperess replied:
See commend on the CMakeLists.txt for resolution.

henrikbrixandersen commented:
> @stephanosio can this merge? It includes bug fixes for the ICM42688 driver along with tests for those fixes.

It includes a lot more than the bug fixes. Why not separate those to another PR, which can be merged for v3.3?

stephanosio commented:
> > @stephanosio can this merge? It includes bug fixes for the ICM42688 driver along with tests for those fixes.
> 
> It includes a lot more than the bug fixes. Why not separate those to another PR, which can be merged for v3.3?

As @henrikbrixandersen mentioned, please put the bug fixes into a separate PR for v3.3.

yperess commented:
> > > @stephanosio can this merge? It includes bug fixes for the ICM42688 driver along with tests for those fixes.
> > 
> > 
> > It includes a lot more than the bug fixes. Why not separate those to another PR, which can be merged for v3.3?
> 
> As @henrikbrixandersen mentioned, please put the bug fixes into a separate PR for v3.3.

I'll just wait for the feature freeze to finish. I saw that the rules for merging are: "bug fixes, cosmetic changes, and tests" which I thought this fits, but if not I'll just wait so I can focus on the next PR that builds on this one.

henrikbrixandersen reviewed (changes requested):
Please address the file naming pointed out earlier.

Commit:
icm42688: fix configure call

The icm42688 introduced a safe configure function to make the
configuration more atomic, but the main driver was never updated
to call it so the new config values were discarded.

Signed-off-by: Yuval Peress <peress@google.com>

henrikbrixandersen reviewed (approved):
EEPROM looks good, thanks.
